src/libraries/System.Security.Cryptography/src/System/Security/Cryptography/X509Certificates/X509Certificate2.cs

```diff
@@ -1166,6 +1167,239 @@ public bool TryExportCertificatePem(Span<char> destination, out int charsWritten
+                    if (hostname.EndsWith('.'))
+                    {
+                        match = match.Slice(0, match.Length - 1);
+
+                        if (match.IsEmpty)
+                        {
+                            return false;
+                        }
+                    }
+
+                    ReadOnlySpan<char> afterFirstDot = default;
+                    int firstDot = match.IndexOf('.');
+
+                    // ".something.example.org" always fails to match.
+                    if (firstDot == 0)
+                    {
+                        return false;
+                    }
+
+                    if (firstDot > 0)
+                    {
+                        afterFirstDot = match.Slice(firstDot + 1);
+                    }
+
+                    foreach (string embedded in san.EnumerateDnsNames())
+                    {
+                        hadAny = true;
+
+                        if (embedded.Length == 0)
+                        {
+                            continue;
+                        }
+
+                        ReadOnlySpan<char> embeddedSpan = embedded;
+
+                        // Convert embedded "something.example.org." to "something.example.org"
+                        if (embedded.EndsWith('.'))
+                        {
+                            embeddedSpan = embeddedSpan.Slice(0, embeddedSpan.Length - 1);
+                        }
+
+                        if (allowWildcards && embeddedSpan.StartsWith("*.") && embeddedSpan.Length > 2)
+                        {
+                            if (embeddedSpan.Slice(2).Equals(afterFirstDot, StringComparison.OrdinalIgnoreCase))
+                            {
+                                return true;
+                            }
+                        }
+                        else if (embeddedSpan.Equals(match, StringComparison.OrdinalIgnoreCase))
+                        {
+                            return true;
+                        }
+                    }
+                }
+
+                if (hadAny)
+                {
+                    return false;
+                }
+            }
+
+            if (allowCommonName)
+            {
+                X500RelativeDistinguishedName? cn = null;
+
+                foreach (X500RelativeDistinguishedName rdn in SubjectName.EnumerateRelativeDistinguishedNames())
+                {
+                    if (rdn.HasMultipleElements)
+                    {
+                        AsnValueReader reader = new AsnValueReader(rdn.RawData.Span, AsnEncodingRules.DER);
+                        // Be lax with the sort order because Windows is
+                        AsnValueReader set = reader.ReadSetOf(skipSortOrderValidation: true);
+
+                        while (set.HasData)
+                        {
+                            AsnValueReader attributeTypeAndValue = set.ReadSequence();
```

Comment:
I think this could use a comment something of the affect of "X500RelativeDistinguishedName ensures the ASN.1 is already valid, so we don't need to check for trailing data"

Assuming I am right about that.